File: src/dr_stone/normalizers.py
from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation

from dr_stone.exceptions import ParseError


def normalize_currency(value: str | None) -> str:
    if not value:
        return "BRL"

    normalized = value.strip().upper()
    if normalized in {"R$", "BRL"}:
        return "BRL"
    return normalized


def normalize_price(value: str | int | float | Decimal) -> Decimal:
    if isinstance(value, Decimal):
        return value.quantize(Decimal("0.01"))
    if isinstance(value, int | float):
        return Decimal(str(value)).quantize(Decimal("0.01"))

    cleaned = re.sub(r"[^\d,.\-]", "", value.strip())
    if not cleaned:
        raise ParseError("Price value is empty after normalization")

    if "," in cleaned and "." in cleaned:
        if cleaned.rfind(",") > cleaned.rfind("."):
            cleaned = cleaned.replace(".", "").replace(",", ".")
        else:
            cleaned = cleaned.replace(",", "")
    elif "," in cleaned:
        cleaned = cleaned.replace(".", "").replace(",", ".")

    try:
        return Decimal(cleaned).quantize(Decimal("0.01"))
    except InvalidOperation as exc:
        raise ParseError(f"Unable to parse price value: {value}") from exc


def normalize_availability(value: str | None) -> tuple[str, bool]:
    if not value:
        return "unknown", False

    normalized = value.strip().lower()
    if normalized.endswith("/instock") or normalized in {"in_stock", "instock"}:
        return "in_stock", True
    if normalized.endswith("/outofstock") or normalized in {"out_of_stock", "outofstock"}:
        return "out_of_stock", False
    if "dispon" in normalized:
        return "in_stock", True
    if "indispon" in normalized or "esgot" in normalized:
        return "out_of_stock", False
    return normalized, False

File: dr-stone-scrapper/src/normalizers.ts
import { ParseError } from "./errors.js";

export function normalizeTitle(value: string): string {
  return value.normalize("NFKC").toLocaleLowerCase().split(/\s+/).filter(Boolean).join(" ");
}

export function titleContainsAllTerms(searchTerms: string[], candidateTitle: string): boolean {
  const normalizedCandidate = normalizeTitle(candidateTitle);
  return searchTerms.every((term) => normalizedCandidate.includes(normalizeTitle(term)));
}

export function normalizeCurrency(value?: string | null): string {
  if (!value) {
    return "BRL";
  }

  const normalized = value.trim().toUpperCase();
  if (normalized === "R$" || normalized === "BRL") {
    return "BRL";
  }

  return normalized;
}

export function normalizePrice(value: string | number): string {
  if (typeof value === "number") {
    return value.toFixed(2);
  }

  const cleaned = value.replace(/[^\d,.-]/g, "");
  if (!cleaned) {
    throw new ParseError("Price value is empty after normalization");
  }

  let normalized = cleaned;
  if (normalized.includes(",") && normalized.includes(".")) {
    normalized =
      normalized.lastIndexOf(",") > normalized.lastIndexOf(".")
        ? normalized.replaceAll(".", "").replace(",", ".")
        : normalized.replaceAll(",", "");
  } else if (normalized.includes(",")) {
    normalized = normalized.replaceAll(".", "").replace(",", ".");
  }

  const parsed = Number(normalized);
  if (Number.isNaN(parsed)) {
    throw new ParseError(`Unable to parse price value: ${value}`);
  }

  return parsed.toFixed(2);
}

export function normalizeAvailability(value?: string | null): [string, boolean] {
  if (!value) {
    return ["unknown", false];
  }

  const normalized = value.trim().toLowerCase();
  if (normalized.endsWith("/instock") || normalized === "in_stock" || normalized === "instock") {
    return ["in_stock", true];
  }

  if (
    normalized.endsWith("/outofstock") ||
    normalized === "out_of_stock" ||
    normalized === "outofstock"
  ) {
    return ["out_of_stock", false];
  }

  if (normalized.includes("indispon") || normalized.includes("esgot")) {
    return ["out_of_stock", false];
  }

  if (normalized.includes("dispon")) {
    return ["in_stock", true];
  }

  return [normalized, false];
}
